pkg/controller/controlplane/actuator.go

```diff
@@ -0,0 +1,148 @@
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/controlplane"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils/chart"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+const (
+	namespacesChartName = "namespaces"
+)
+
+// NewActuator creates a new Actuator that acts upon and updates the status of ControlPlane resources.
+func NewActuator(
+	a controlplane.Actuator,
+	additionalCharts []*chart.Chart,
+	chartRendererFactory extensionscontroller.ChartRendererFactory,
+) controlplane.Actuator {
+	return &actuator{
+		Actuator:             a,
+		additionalCharts:     additionalCharts,
+		chartRendererFactory: chartRendererFactory,
+	}
+}
+
+// actuator is a controlplane.Actuator that acts upon and updates the status of ControlPlane resources.
+// This one runs the normal control plane activities, and then also applies the additional charts, if any
+// chart.
+type actuator struct {
+	controlplane.Actuator
+	client                         client.Client
+	chartRendererFactory           extensionscontroller.ChartRendererFactory
+	additionalChartsValuesProvider valuesProvider
+	additionalCharts               []*chart.Chart
+}
+
+// InjectFunc enables injecting Kubernetes dependencies into actuator's dependencies.
+func (a *actuator) InjectFunc(f inject.Func) error {
+	return f(a.Actuator)
+}
+
+// InjectClient injects the given client into the valuesProvider.
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+// Reconcile reconcile the core controlplane chart, and then any additional charts.
+func (a *actuator) Reconcile(
+	ctx context.Context,
+	cp *extensionsv1alpha1.ControlPlane,
+	cluster *extensionscontroller.Cluster,
+) (bool, error) {
+	// Call Reconcile on the composed Actuator
+	requeue, err := a.Actuator.Reconcile(ctx, cp, cluster)
+	if err != nil {
+		return false, err
+	}
+
+	ivector := imagevector.ImageVector()
+	version := cluster.Shoot.Spec.Kubernetes.Version
+
+	// Create shoot chart renderer
+	chartRenderer, err := a.chartRendererFactory.NewChartRendererForShoot(version)
+	if err != nil {
+		return false, fmt.Errorf("could not create chart renderer for shoot '%s': %w", cp.Namespace, err)
+	}
+
+	// extract any subchart named "namespaces" if it exists and pre-process it
+	for _, subc := range a.additionalCharts {
+		if subc.Name == namespacesChartName {
+			// Render each shoot additional components chart and create a managed resource
+			if err := managedresources.RenderChartAndCreate(ctx, cp.Namespace, subc.Name, false, a.client, chartRenderer, subc, nil, ivector, "", version, true, false); err != nil {
+				return false, err
+			}
+		}
+	}
```

Comment:
With the latest changes this is simply not needed and can be removed. A managed resource for the namespaces chart will be created in the for loop below. Note that the order in which the managed resources are created doesn't matter - as I pointed out several times, they will be applied in parallel by the GRM and it's impossible to control any dependencies between them right now. If this turns out to be a problem, we might need to enhance GRM a bit.

Reply:
Ah, I thought I had removed this. Good catch.